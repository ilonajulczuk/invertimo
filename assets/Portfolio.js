import React, { Suspense } from 'react';

import PropTypes from 'prop-types';
import {
    Switch,
    Route,
    NavLink,
    Redirect,
} from "react-router-dom";

import Icon from '@mui/material/Icon';
import Button from '@mui/material/Button';

import makeStyles from '@mui/styles/makeStyles';
import Typography from '@mui/material/Typography';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';

import './portfolio.css';
import { Events } from './Events.js';

import { Header } from './Header.js';
import { APIClient } from './api_utils.js';
import { toSymbol } from './currencies.js';
import { ErrorBoundary } from './error_utils.js';
import { Accounts } from './Accounts.js';


const PositionList = React.lazy(() => import('./PositionList'));
const Onboarding = React.lazy(() => import('./Onboarding'));
const Transactions = React.lazy(() => import('./Transactions'));
const AccountValue = React.lazy(() => import('./AccountValue'));


const useStyles = makeStyles({
    root: {
        minWidth: 275,
    },
    title: {
        fontSize: 18,
    },
});


function AccountStats(props) {

    const classes = useStyles();
    const account = props.account;

    let totalCash = Number(account.balance);
    let currencySymbol = toSymbol(account.currency);

    const assetValues = props.positions.map(position => {
        // Exchange rate is null if account and position trading currency are the same.
        let latest_exchange_rate = position.latest_exchange_rate ? position.latest_exchange_rate : 1;
        return position.quantity * position.latest_price * latest_exchange_rate;
    });
    const totalAssetValue = Math.round(
        assetValues.reduce((sum, current) => sum + current, 0) * 100) / 100;

    return (
        <Card className={classes.root} variant="outlined">
            <CardContent>
                <Typography
                    className={classes.title}
                    variant="h5" component="h5" gutterBottom>
                    Account <a href="#/accounts">{account.nickname}</a>
                </Typography>
                <Typography variant="body1" component="p">
                    Total Cash: {totalCash} {currencySymbol}
                </Typography>
                <Typography variant="body1" component="p">
                    Total Assets: {totalAssetValue} {currencySymbol}
                </Typography>
                <Typography variant="body1" component="p">
                    Portfolio Value: {Math.round((totalCash + totalAssetValue) * 100) / 100} {currencySymbol}
                </Typography>
            </CardContent>

        </Card>);
}


AccountStats.propTypes = {
    positions: PropTypes.array.isRequired,
    account: PropTypes.object.isRequired,
};


export function divideByAccount(accounts, positions) {
    const positionsByAccount = new Map();
    accounts.forEach(account => positionsByAccount.set(account.id, { account: account, positions: [] }));
    positions.forEach(position => {
        const accountEntry = positionsByAccount.get(position.account);
        if (accountEntry) {
            accountEntry.positions.push(position);
        }

    });
    return positionsByAccount;
}


export function PortfolioOverview(props) {
    let accounts = props.accounts;

    const positionsByAccount = divideByAccount(accounts, props.positions);
    const accountStatsEntries = Array.from(positionsByAccount).map(([, entry]) => {
        return (
            <AccountStats
                key={entry.account.id}
                account={entry.account}
                positions={entry.positions}
            />
        );
    });
    return (
        <div className="portfolio-overview">
            <div className="quick-actions">
                <span className="card-label">Quick actions</span>
                <div>
                    <Button
                        href="#/transactions/record"
                        variant="contained"
                        color="secondary"
                    >
                        <Icon>create</Icon>
                    Record transaction
                    </Button>
                    <Button
                        href="#/events/record_dividend"
                        variant="contained"
                        color="secondary"
                    >
                        <Icon>paid</Icon>
                    Record dividend
                    </Button>
                    <Button
                        href="#/events/record_transfer"
                        variant="contained"
                        color="secondary"
                    >
                        <Icon>sync_alt</Icon>
                    Record transfer
                    </Button>
                    <Button
                        href="#/accounts"
                        variant="contained"
                        color="primary"
                    >
                        Manage accounts
                    </Button>
                    <Button
                        href="#/start/investment_accounts"
                        variant="contained"
                        color="primary"
                    >
                        Quickstart
                    </Button>
                </div>

            </div>
            <div>
                <h2>Account stats</h2>
            </div>

            <div className="account-stats-list" >
                {accountStatsEntries}

            </div>


        </div>
    );
}

PortfolioOverview.propTypes = {
    positions: PropTypes.array.isRequired,
    accounts: PropTypes.array.isRequired,
};


export default class Portfolio extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            "positions": null,
            "positionDetails": new Map(),
            "accounts": null,
            "transactions": null,
            "events": null,
            "accountValues": new Map(),
        };
        this.apiClient = new APIClient('./api');
        this.getPositionDetail = this.getPositionDetail.bind(this);
        this.handleAddAccount = this.handleAddAccount.bind(this);
        this.handleAddTransaction = this.handleAddTransaction.bind(this);
        this.handleDeleteTransaction = this.handleDeleteTransaction.bind(this);
        this.handleCorrectTransaction = this.handleCorrectTransaction.bind(this);
        this.handleAddEvent = this.handleAddEvent.bind(this);
        this.handleDeleteEvent = this.handleDeleteEvent.bind(this);
        this.handleDeleteAccount = this.handleDeleteAccount.bind(this);
    }

    async handleAddAccount(accountData) {
        const data = {
            nickname: accountData.name,
            currency: accountData.currency,
            description: "",
        };
        let result = await this.apiClient.createAccount(data);
        // Reload all the data, e.g. accounts, positions, etc.
        if (result.ok) {
            this.refreshFromServer();
        }
        return result;
    }

    async handleAddTransaction(data) {
        let result;
        if (data.asset) {
            result = await this.apiClient.addTransaction(data);
        } else {
            result = await this.apiClient.addTransactionWithCustomAsset(data);
        }
        // Reload all the data, e.g. accounts, positions, etc.
        if (result.ok) {
            this.refreshFromServer();
        }
        return result;
    }

    async handleDeleteTransaction(transactionId) {
        let result = await this.apiClient.deleteTransaction(transactionId);
        // Reload all the data, e.g. accounts, positions, etc.
        if (result.ok) {
            this.refreshFromServer();
        }
        return result;
    }

    async handleCorrectTransaction(transactionId, update) {
        let result = await this.apiClient.correctTransaction(transactionId, update);
        if (result.ok) {
            this.refreshFromServer();
        }
        return result;
    }

    async getPositionDetail(positionId) {
        if (this.state.positionDetails.has(positionId)) {
            return this.state.positionDetails.get(positionId);
        }
        return this.fetchAndSetPositionDetail(positionId);

    }

    async fetchAndSetPositionDetail(positionId) {
        try {
            let positionData = await this.apiClient.getPositionDetail(positionId);
            let positionDetails = this.state.positionDetails;
            positionDetails.set(positionId, positionData);
            this.setState({
                "positionDetails": positionDetails,
            });
            return positionData;
        }
        catch (error) {
            alert(error);
            return null;
        }

    }

    async handleAddEvent(data) {
        let result = await this.apiClient.addEvent(data);
        if (result.ok) {
            this.refreshFromServer();
        }
        return result;
    }

    async handleDeleteEvent(eventId) {
        let result = await this.apiClient.deleteEvent(eventId);
        // Reload all the data, e.g. accounts, positions, etc.
        this.refreshFromServer();
        return result;
    }

    async handleDeleteAccount(accountId) {
        let result = await this.apiClient.deleteAccount(accountId);
        // Reload all the data, e.g. accounts, positions, etc.
        this.refreshFromServer({deleteAccount: Number(accountId)});

        return result;
    }

    async componentDidMount() {
        this.refreshFromServer();
    }

    refreshFromServer({deleteAccount = null} = {}) {
        this.apiClient.getAccounts().then(accounts => {
            this.setState({ "accounts": accounts });
            if (deleteAccount) {
                let accountValues = this.state.accountValues;
                accountValues.delete(deleteAccount);
                this.setState({
                    "accountValues": accountValues,
                });
            }
            accounts.forEach(account => {
                this.apiClient.getAccountDetail(account.id, 4 * 365).then(account => {
                    let accountValues = this.state.accountValues;
                    accountValues.set(account.id, account);
                    this.setState({
                        "accountValues": accountValues,
                    });
                });
            });
        }
        );
        this.apiClient.getPositions().then(
            positions => {
                this.setState(
                    {
                        "positions": positions,
                        "positionDetails": new Map(),
                    });
                this.apiClient.getTransactions().then(
                    transactions => {
                        this.setState({ "transactions": transactions });
                    });
            });
        this.apiClient.getEvents().then(
            events => {
                this.setState({ "events": events });
            }
        );
    }

    render() {
        const userEmail = JSON.parse(document.getElementById('userEmail').textContent);


        const navBar = <nav className="sidenav">
            <ul>
                <li>
                    <NavLink to="/" exact={true}>Home</NavLink>
                </li>
                <li>
                    <NavLink to="/positions">Positions</NavLink>
                </li>
                <li>
                    <NavLink to="/transactions">Transactions</NavLink>
                </li>
                <li>
                    <NavLink to="/events">Events</NavLink>
                </li>
                <li>
                    <NavLink to="/accounts">Accounts</NavLink>
                </li>
            </ul>
        </nav>;

        // If there are no accounts loaded yet, there isn't much to show.
        // TODO: start displaying sth useful even if the big part is still loading.
        if (this.state.accounts === null) {
            return (<div className="main-grid">
                <Header email={userEmail} />
                {navBar}
                <div className="main-content">
                    <div>
                        <h2>Loading...</h2>
                        <p>Loading basic info...</p>
                    </div>
                </div>
            </div>);
        }
        if (this.state.positions === null || this.state.accountValues.size !== this.state.accounts.length) {
            return (<div className="main-grid">
                <Header email={userEmail} />
                {navBar}
                <div className="main-content">
                    <h2>Loading...</h2>
                    <div>
                        <p>Loading graphs and counting your investments...</p>
                    </div>
                </div>
            </div>);
        }

        // If the accounts are loaded, but there is nothing there, start an onboarding wizard.
        const newUser = this.state.accounts.length == 0;

        let numTransactions = 0;
        if (!newUser) {
            for (let account of this.state.accounts) {
                numTransactions += account.transactions_count;
            }
        }
        const noTransactions = numTransactions == 0;

        let accountValues = this.state.accounts.filter(account =>
            this.state.accountValues.get(account.id)).map((account) => {

                let accountDetail = this.state.accountValues.get(account.id);
                let values = [];
                if (accountDetail) {
                    values = accountDetail.values;
                }
                return (
                    <AccountValue key={account.id} account={account}
                        positions={this.state.positions} values={values} />
                );
            });

        let redirectOrDisplay;
        if (newUser) {
            redirectOrDisplay = <Redirect to="/start/investment_accounts" />;
        } else if (noTransactions) {
            redirectOrDisplay = <Redirect to="/start/transactions_intro" />;
        } else {
            redirectOrDisplay = (
                <div>
                    <ErrorBoundary>
                        <PortfolioOverview positions={this.state.positions} accounts={this.state.accounts} />
                        <Suspense fallback={<div>Loading graphs...</div>}>
                            {accountValues}
                        </Suspense>
                    </ErrorBoundary>
                </div>);
        }
        let maybeTransactions = <h2>Loading transactions...</h2>;
        if (this.state.transactions !== null) {
            maybeTransactions = <Transactions transactions={this.state.transactions}
                handleAddTransaction={this.handleAddTransaction}
                handleDeleteTransaction={this.handleDeleteTransaction}
                handleCorrectTransaction={this.handleCorrectTransaction}
                positions={this.state.positions}
                accounts={this.state.accounts} />;
        }
        let maybeEvents = <h2>Loading events...</h2>;
        if (this.state.events !== null) {
            maybeEvents = (<Events
                accounts={this.state.accounts}
                events={this.state.events} positions={this.state.positions}
                handleAddEvent={this.handleAddEvent}
                handleDeleteEvent={this.handleDeleteEvent}
            />
            );
        }
        return (
            <div className="main-grid">
                <Header email={userEmail} />

                {navBar}
                <div className="main-content">
                    <Suspense fallback={<div><h2>Loading...</h2>
                        <p>Checking how the market is doing...</p></div>}>
                        <Switch>
                            <Route path="/transactions">
                                <ErrorBoundary>
                                    {maybeTransactions}
                                </ErrorBoundary>
                            </Route>
                            <Route path="/positions">
                                <ErrorBoundary>

                                    <PositionList positions={this.state.positions}
                                        accounts={this.state.accounts} getPositionDetail={this.getPositionDetail} />

                                </ErrorBoundary>
                            </Route>
                            <Route path="/events">
                                <ErrorBoundary>
                                    {maybeEvents}
                                </ErrorBoundary>
                            </Route>
                            <Route path="/accounts">
                                <ErrorBoundary>
                                    <Accounts
                                        accounts={this.state.accounts}
                                        handleAddAccount={this.handleAddAccount}
                                        handleDeleteAccount={this.handleDeleteAccount} />
                                </ErrorBoundary>
                            </Route>
                            <Route path="/start/:stepName">
                                <Onboarding accounts={this.state.accounts}
                                    handleAddAccount={this.handleAddAccount}
                                    handleAddTransaction={this.handleAddTransaction}
                                    transactions={this.state.transactions}
                                />

                            </Route>
                            <Route path="/">
                                {redirectOrDisplay}
                            </Route>

                        </Switch>
                    </Suspense>
                </div>
            </div>
        );

    }
}