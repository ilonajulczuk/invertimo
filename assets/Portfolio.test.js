import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import pretty from "pretty";

// The import below is necessary for async/await to work.
// eslint-disable-next-line no-unused-vars
import regeneratorRuntime from "regenerator-runtime";
import Portfolio from "./Portfolio.js";
import { PortfolioOverview, divideByAccount } from "./Portfolio.js";
import { MemoryRouter } from "react-router-dom";


const _ACCOUNT_POSITIONS = [
  {
    id: 91,
    account: 3,
    security: {
      id: 1,
      isin: "US2546871060",
      symbol: "DIS",
      name: "The Walt Disney Company",
      exchange: {
        id: 132,
        name: "USA Stocks",
      },
      currency: "USD",
      country: "USA",
    },
    quantity: "35.00",
    latest_price: "176.24",
    latest_price_date: "2021-06-03",
    latest_exchange_rate: "0.82470000",
  },
  {
    id: 93,
    account: 3,
    security: {
      id: 3,
      isin: "US4781601046",
      symbol: "JNJ",
      name: "Johnson & Johnson",
      exchange: {
        id: 132,
        name: "USA Stocks",
      },
      currency: "USD",
      country: "USA",
    },
    quantity: "29.00",
    latest_price: "166.04",
    latest_price_date: "2021-06-03",
    latest_exchange_rate: "0.82470000",
  },
  {
    id: 96,
    account: 3,
    security: {
      id: 6,
      isin: "IE00BF4RFH31",
      symbol: "IUSN",
      name: "iShares MSCI World Small Cap UCITS ETF USD (Acc)",
      exchange: {
        id: 143,
        name: "XETRA Exchange",
      },
      currency: "EUR",
      country: "Germany",
    },
    quantity: "2775.00",
    latest_price: "6.06",
    latest_price_date: "2021-06-04",
    latest_exchange_rate: null,
  },
];

const _ACCOUNT_POSITIONS_USD = [
  {
    id: 121,
    account: 4,
    security: {
      id: 1,
      isin: "US2546871060",
      symbol: "DIS",
      name: "The Walt Disney Company",
      exchange: {
        id: 132,
        name: "USA Stocks",
      },
      currency: "USD",
      country: "USA",
    },
    quantity: "16.00",
    latest_price: "176.24",
    latest_price_date: "2021-06-03",
    latest_exchange_rate: null,
  },
  {
    id: 122,
    account: 4,
    security: {
      id: 2,
      isin: "US0378331005",
      symbol: "AAPL",
      name: "Apple Inc",
      exchange: {
        id: 132,
        name: "USA Stocks",
      },
      currency: "USD",
      country: "USA",
    },
    quantity: "18.00",
    latest_price: "123.54",
    latest_price_date: "2021-06-03",
    latest_exchange_rate: null,
  },
];


const _ACCOUNT_TRANSACTIONS = [
  {
    "id": 666,
    "executed_at": "2021-04-07T20:25:00Z",
    "last_modified": "2021-06-06T17:14:28.728056Z",
    "position": {
      "id": 121,
      "account": 4,
      "security": {
        "id": 1,
        "isin": "US2546871060",
        "symbol": "DIS",
        "name": "The Walt Disney Company",
        "exchange": {
          "id": 132,
          "name": "USA Stocks"
        },
        "currency": "USD",
        "country": "USA"
      },
      "quantity": "16.00"
    },
    "quantity": "16.00",
    "price": "188.12",
    "transaction_costs": "-0.55",
    "local_value": "-3009.92",
    "value_in_account_currency": "-2533.39",
    "total_in_account_currency": "-2533.94",
    "order_id": "420be97a-265e-4b69-b7a7-acfad1ba8940"
  },
  {
    "id": 531,
    "executed_at": "2021-04-07T20:25:00Z",
    "last_modified": "2021-06-06T15:07:31.366491Z",
    "position": {
      "id": 91,
      "account": 3,
      "security": {
        "id": 1,
        "isin": "US2546871060",
        "symbol": "DIS",
        "name": "The Walt Disney Company",
        "exchange": {
          "id": 132,
          "name": "USA Stocks"
        },
        "currency": "USD",
        "country": "USA"
      },
      "quantity": "35.00"
    },
    "quantity": "16.00",
    "price": "188.12",
    "transaction_costs": "-0.55",
    "local_value": "-3009.92",
    "value_in_account_currency": "-2533.39",
    "total_in_account_currency": "-2533.94",
    "order_id": "420be97a-265e-4b69-b7a7-acfad1ba8940"
  },
  {
    "id": 532,
    "executed_at": "2021-04-07T20:23:00Z",
    "last_modified": "2021-06-06T15:07:31.373143Z",
    "position": {
      "id": 92,
      "account": 3,
      "security": {
        "id": 2,
        "isin": "US0378331005",
        "symbol": "AAPL",
        "name": "Apple Inc",
        "exchange": {
          "id": 132,
          "name": "USA Stocks"
        },
        "currency": "USD",
        "country": "USA"
      },
      "quantity": "53.00"
    },
    "quantity": "18.00",
    "price": "127.64",
    "transaction_costs": "-0.56",
    "local_value": "-2297.52",
    "value_in_account_currency": "-1933.78",
    "total_in_account_currency": "-1934.34",
    "order_id": "5e9d4dd0-6d3a-4ca2-aba5-356ac49e7ab4"
  },
];
jest.mock("./api_utils", () => {
  // Works and lets you check for constructor calls:
  return {
    APIClient: jest.fn().mockImplementation(() => {
      return {
        getAccounts: () => {
          let accounts = [
            {
              id: 3,
              currency: "EUR",
              nickname: "test account",
              description: "",
              balance: "-106584.76000",
              last_modified: "2021-05-03T14:20:29.732353Z",
              positions_count: 30,
              transactions_count: 135,
            },
          ];
          return Promise.resolve(accounts);
        },
        getPositions: () => {
          return Promise.resolve(_ACCOUNT_POSITIONS);
        },
        getPositionDetail: () => { },
        getTransactions: () => {

          return Promise.resolve(_ACCOUNT_TRANSACTIONS);
        },
      };
    }),
  };
});

let container = null;

describe('portfolio tests', () => {

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });


  it("shows portfolio overview with one account", async () => {
    expect.hasAssertions();
    const accounts = [
      {
        id: 3,
        currency: "EUR",
        nickname: "test account",
        description: "",
        balance: "-16584.76000",
        last_modified: "2021-05-03T14:20:29.732353Z",
        positions_count: 3,
        transactions_count: 15,
      },
    ];
    const positions = _ACCOUNT_POSITIONS;
    await act(async () => {
      render(
        <PortfolioOverview accounts={accounts} positions={positions} />,
        container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("divides positions by account correctly", () => {
    const accounts = [
      {
        id: 3,
        currency: "EUR",
        nickname: "test account",
        description: "",
        balance: "-16584.76000",
        last_modified: "2021-05-03T14:20:29.732353Z",
        positions_count: 3,
        transactions_count: 15,
      },
      {
        id: 4,
        currency: "USD",
        nickname: "second account",
        description: "",
        balance: "161",
        last_modified: "2021-05-03T14:20:29.732353Z",
        positions_count: 2,
        transactions_count: 5,
      },
    ];

    const positions = _ACCOUNT_POSITIONS.concat(_ACCOUNT_POSITIONS_USD);
    const positionsByAccount = divideByAccount(accounts, positions);
    expect(positionsByAccount).toHaveLength(2);
    expect(positionsByAccount[0].account.nickname).toBe("test account");
    expect(positionsByAccount[0].positions).toHaveLength(3);
    expect(positionsByAccount[1].account.nickname).toBe("second account");
    expect(positionsByAccount[1].positions).toHaveLength(2);
  });

  it("shows portfolio overview with two accounts", async () => {
    expect.hasAssertions();
    const accounts = [
      {
        id: 3,
        currency: "EUR",
        nickname: "test account",
        description: "",
        balance: "-16584.76000",
        last_modified: "2021-05-03T14:20:29.732353Z",
        positions_count: 3,
        transactions_count: 15,
      },
      {
        id: 4,
        currency: "USD",
        nickname: "second account",
        description: "",
        balance: "161",
        last_modified: "2021-05-03T14:20:29.732353Z",
        positions_count: 2,
        transactions_count: 5,
      },
    ];

    const positions = _ACCOUNT_POSITIONS.concat(_ACCOUNT_POSITIONS_USD);
    await act(async () => {
      render(
        <PortfolioOverview accounts={accounts} positions={positions} />,
        container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("shows portfolio overview without any accounts", async () => {
    expect.hasAssertions();
    const accounts = [];

    const positions = [];
    await act(async () => {
      render(
        <PortfolioOverview accounts={accounts} positions={positions} />,
        container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("renders portfolio value", async () => {
    expect.hasAssertions();
    await act(async () => {
      render(<MemoryRouter initialEntries={['/positions']}>
        <Portfolio />
      </MemoryRouter>, container);
    });
    expect(pretty(container.innerHTML)).toContain('The Walt Disney Company');
  });

});