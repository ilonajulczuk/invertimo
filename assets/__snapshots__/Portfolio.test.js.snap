// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders portfolio value 1`] = `
"<div class=\\"main-grid\\">
  <nav class=\\"sidenav\\">
    <ul>
      <li><a href=\\"/\\">Home</a></li>
      <li><a href=\\"/transactions\\">Transactions</a></li>
      <li><a aria-current=\\"page\\" class=\\"active\\" href=\\"/positions\\">Positions</a></li>
    </ul>
  </nav>
  <div class=\\"main-content\\">
    <h1>Portfolio</h1>
    <div class=\\"portfolio-overview\\">
      <div class=\\"portfolio-overview-card card\\"><span class=\\"card-label\\">At a glance</span>
        <ul class=\\"portfolio-stats-list\\">
          <li class=\\"portfolio-stats-account-name\\">In <a href=\\"\\">test account</a> account:</li>
          <li>Total Cash: -106584.76 €</li>
          <li>Total Assets: 25874.64 €</li>
          <li>Portfolio Value: -80710.12 €</li>
        </ul>
      </div>
      <div class=\\"card\\"><span class=\\"card-label\\">Assets</span>
        <p>30 <a href=\\"\\">Positions</a> in 1 <a href=\\"\\"> Account</a></p><a class=\\"button\\">See all Positions</a><a class=\\"button\\">Manage accounts</a>
      </div>
      <div class=\\"card\\"><span class=\\"card-label\\">Events</span>
        <div>135 <a href=\\"\\">Transactions</a></div>
        <div>? <a href=\\"\\">Account Events</a></div><a class=\\"button\\">Manage transactions</a><a class=\\"button\\">Manage events</a>
      </div>
    </div>
    <div>
      <h2>Positions</h2>
      <ul class=\\"position-list\\">
        <li class=\\"position-list-header\\">
          <ul class=\\"position-list-fields\\">
            <li class=\\"position-list-fields-product\\">Product</li>
            <li>Exchange</li>
            <li>Quantity</li>
            <li>Price</li>
            <li>Value</li>
          </ul>
        </li>
        <li>
          <div class=\\"position-card \\">
            <div class=\\"position-name\\"><span class=\\"card-label\\">US2546871060</span><span class=\\"position-symbol\\">DIS</span><span>The Walt Disney Company</span></div>
            <div>USA Stocks</div>
            <div>35.00</div>
            <div><span class=\\"card-label\\">As of 2021-06-03</span>176.24 USD</div>
            <div><span>6168.4 USD</span><span>5087.08 EUR</span></div>
          </div>
        </li>
        <li>
          <div class=\\"position-card \\">
            <div class=\\"position-name\\"><span class=\\"card-label\\">US4781601046</span><span class=\\"position-symbol\\">JNJ</span><span>Johnson &amp; Johnson</span></div>
            <div>USA Stocks</div>
            <div>29.00</div>
            <div><span class=\\"card-label\\">As of 2021-06-03</span>166.04 USD</div>
            <div><span>4815.16 USD</span><span>3971.06 EUR</span></div>
          </div>
        </li>
        <li>
          <div class=\\"position-card \\">
            <div class=\\"position-name\\"><span class=\\"card-label\\">IE00BF4RFH31</span><span class=\\"position-symbol\\">IUSN</span><span>iShares MSCI World Small Cap UCITS ETF USD (Acc)</span></div>
            <div>XETRA Exchange</div>
            <div>2775.00</div>
            <div><span class=\\"card-label\\">As of 2021-06-04</span>6.06 EUR</div>
            <div><span>16816.5 EUR</span><span></span></div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>"
`;

exports[`shows portfolio overview with one account 1`] = `
"<div class=\\"portfolio-overview\\">
  <div class=\\"portfolio-overview-card card\\"><span class=\\"card-label\\">At a glance</span>
    <ul class=\\"portfolio-stats-list\\">
      <li class=\\"portfolio-stats-account-name\\">In <a href=\\"\\">test account</a> account:</li>
      <li>Total Cash: -16584.76 €</li>
      <li>Total Assets: 25874.64 €</li>
      <li>Portfolio Value: 9289.88 €</li>
    </ul>
  </div>
  <div class=\\"card\\"><span class=\\"card-label\\">Assets</span>
    <p>3 <a href=\\"\\">Positions</a> in 1 <a href=\\"\\"> Account</a></p><a class=\\"button\\">See all Positions</a><a class=\\"button\\">Manage accounts</a>
  </div>
  <div class=\\"card\\"><span class=\\"card-label\\">Events</span>
    <div>15 <a href=\\"\\">Transactions</a></div>
    <div>? <a href=\\"\\">Account Events</a></div><a class=\\"button\\">Manage transactions</a><a class=\\"button\\">Manage events</a>
  </div>
</div>"
`;

exports[`shows portfolio overview with two accounts 1`] = `
"<div class=\\"portfolio-overview\\">
  <div class=\\"portfolio-overview-card card\\"><span class=\\"card-label\\">At a glance</span>
    <ul class=\\"portfolio-stats-list\\">
      <li class=\\"portfolio-stats-account-name\\">In <a href=\\"\\">test account</a> account:</li>
      <li>Total Cash: -16584.76 €</li>
      <li>Total Assets: 25874.64 €</li>
      <li>Portfolio Value: 9289.88 €</li>
    </ul>
    <ul class=\\"portfolio-stats-list\\">
      <li class=\\"portfolio-stats-account-name\\">In <a href=\\"\\">second account</a> account:</li>
      <li>Total Cash: 161 $</li>
      <li>Total Assets: 5043.56 $</li>
      <li>Portfolio Value: 5204.56 $</li>
    </ul>
  </div>
  <div class=\\"card\\"><span class=\\"card-label\\">Assets</span>
    <p>5 <a href=\\"\\">Positions</a> in 2 <a href=\\"\\"> Accounts</a></p><a class=\\"button\\">See all Positions</a><a class=\\"button\\">Manage accounts</a>
  </div>
  <div class=\\"card\\"><span class=\\"card-label\\">Events</span>
    <div>20 <a href=\\"\\">Transactions</a></div>
    <div>? <a href=\\"\\">Account Events</a></div><a class=\\"button\\">Manage transactions</a><a class=\\"button\\">Manage events</a>
  </div>
</div>"
`;

exports[`shows portfolio overview without any accounts 1`] = `
"<div class=\\"portfolio-overview\\">
  <div class=\\"portfolio-overview-card card\\"><span class=\\"card-label\\">At a glance</span></div>
  <div class=\\"card\\"><span class=\\"card-label\\">Assets</span>
    <p>0 <a href=\\"\\">Positions</a> in 0 <a href=\\"\\"> Account</a></p><a class=\\"button\\">See all Positions</a><a class=\\"button\\">Manage accounts</a>
  </div>
  <div class=\\"card\\"><span class=\\"card-label\\">Events</span>
    <div>0 <a href=\\"\\">Transactions</a></div>
    <div>? <a href=\\"\\">Account Events</a></div><a class=\\"button\\">Manage transactions</a><a class=\\"button\\">Manage events</a>
  </div>
</div>"
`;
